lib/cache/index.ts

```diff
@@ -0,0 +1,69 @@
+import { CacheOptions, DEFAULT_OPTIONS } from './CacheOptions';
+import { ExpirationFactory } from './expirations/factory';
+import { StorageFactory } from './storages/factory';
+import { CacheFactory } from './caches/factory';
+import { HashService } from '../utils/hash';
+import { CacheProviderFactory } from './cacheProvider/factory';
+
+export { CacheOptions };
+
+type TimeoutCacheOptions = CacheOptions & { timeout: number; };
+
+/**
+ * Caches the result of a method.
+ * @param timeout cache timeout in milliseconds.
+ * @param options (optional) caching options.
+ */
+export function cache(timeout: number): MethodDecorator;
+export function cache(options: TimeoutCacheOptions): MethodDecorator;
+export function cache(timeout: number, options?: CacheOptions): MethodDecorator;
+export function cache(
+  timeoutOrOptions: number | TimeoutCacheOptions,
+  optionsOrVoid: CacheOptions = DEFAULT_OPTIONS,
+): MethodDecorator {
+
+  const { timeout, options } = parseParameters(timeoutOrOptions, optionsOrVoid);
+
+  const hashService = new HashService();
+  const expirationFactory = new ExpirationFactory(timeout, options.expiration);
+  const storageFactory = new StorageFactory(options.size, options.storage);
+  const cacheFactory = new CacheFactory(hashService, expirationFactory, storageFactory);
+  const cacheProvider = new CacheProviderFactory(options.scope, cacheFactory).create();
```

Comment:
In the end only the `cacheProvider` is used. To make the decorator code cleaner, please move above lines in a separate function.